clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java

```diff
@@ -790,6 +812,197 @@ public void testOnSubscriptionUpdatedTransitionsToJoiningOnlyIfNotInGroup() {
         verify(membershipManager, never()).transitionToJoining();
     }
 
+    @Test
+    public void testListenerCallbacksBasic() {
```

Comment:
Will do.

Reply:
@dajac @lianetm—I'm seeing an issue when writing a test that validates failures for `onPartitionsRevoked()`. If an `onPartitionsRevoked()` callback throws an exception, is the intention that we still run the `onPartitionsAssigned()` callback regardless?

Reply:
Yes, that's the intention to keep the behaviour of the legacy coordinator, where `onPartitionsAssigned` is always invoked, even if the `onPartitionsRevoked` fails.

Reply:
Hum... @lianetm In this case, it seems that the reconciliation is not failed when only `onPartitionsRevoked` fails. Is this right? It may be better to just fail the reconciliation when `onPartitionsRevoked` fails. What do you think?

Reply:
I wonder if we should just fail the reconciliation process when `onPartitionsRevoked` fails. What do you think, @lianetm?

Reply:
@lianet said:

> When `onPartitionsRevoked` fails, we intentionally invoke the `onPartitionsAssigned` anyways as I was mentioning above (experience the legacy consumer provides), but we should end up failing the reconciliation as it happens when `onPartitionsAssign` fails. Makes sense @kirktrue ?

This isn't what I'm seeing. In my new `MembershipManagerTest` `testListenerCallbacksThrowsErrorOnPartitionsRevoked`, if an exception is thrown from `onPartitionsRevoked()`, the reconciliation step for assignment is skipped completely such that `onPartitionsAssigned()` is not called. Is that a bug or no?

Reply:
Actually I was missing the fact that the [thenCompose](https://github.com/apache/kafka/blob/bc4e2aaf3381297fbe2e947acac3ed692c7e7da0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L752) after the revocation is probably the one short-circuiting, so we do get the behaviour we want, of just aborting the reconciliation. So we can leave it like it is and enable the test. 

As a follow-up I would make the logic clearer, by having both callbacks leading to the failed reconciliation in the same way, but we can leave that out of this PR. I can handle that when I unify the callbacks with [KAFKA-15835](https://issues.apache.org/jira/browse/KAFKA-15835)

Reply:
Yes, the 'assigned' block that lives inside `thenCompose()` is never called if the parent `Future` was completed exceptionally.

If I'm reading this correctly, we want to change the existing behavior for the `AsyncKafkaConsumer` to _not_ call `onPartitionsAssigned()` if `onPartitionsRevoked()` throws an exception. I'll update my tests accordingly.

That said, I imagine one or more of the integration tests will fail with this change, so they will need to remain enabled only for `GENERIC`.

Reply:
If we cannot reuse the existing integration tests, we will need to add new ones. We could do it in a separate PR though.

Reply:
I've updated the `MembershipManagerImplTest` to include these latest changes. Thanks for clarifying the desired behavior.